MC-35509: Updated the software update guide module with the safe upgrade tool documentation (#8314)
magento/devdocs: src/safe-upgrade-tool/developer.md

---
group: software-update-guide
title: Developer information
functional_areas:
  - Upgrade
---

> This topic contains information for developers who want to know more technical information about the Magento Safe Upgrade Tool (SUT). It is focused on developers who work closely with the Magento source code. You can use this knowledge to customize the SUT's components.

Reviewer: Could we add "(ALPHA)" next to Safe Upgrade Tool ?

## M-Ray integration

> M-Ray integration is an internal integration solution.

Reviewer 2: I'm still not sure what M-Ray is. If it's internal, will the public have access to it?

Author: I'll chase that down, but I think it is an internal developed solution, but devs can access it

Reviewer 3: Initially M-Ray along with SUT was supposed to be available only to Commerce customers. Has this changed and approved by product organization?

Reviewer: @paliarush SUT will be limited to Commerce merchants as you pointed out there is no change on this. 
I think we should talk more about Magento API index instead of M-Ray, as M-Ray is an internal tool that won't be used by SUT users but the output (Magento API index) of it will. @paliarush what do you think?

Author: for the moment, changed it to Magento API index

> M-Ray integration is an internal integration solution.

Reviewer 3: ```suggestion
Replace with brief overview of M-Ray and its purpose.
```

The integration with M-Ray is done through:

`Sut\Domain\MRay\MRayInterface`

> Its implementation is through the `config/services.yaml` file. Its value decides where the response of methods `api()` and `modules()` comes from.

Reviewer 2: ```suggestion
It is implemented through the `config/services.yaml` file. Its value decides where the response of methods `api()` and `modules()` comes from.
```

*  `dev/mray_mock_files/api.json`
*  `dev/mray_mock_files/modules.json`

{:.bs-callout-info}
When you make changes to the `services.yaml` file, delete the `var/cache/` folder to correctly apply them.

## Unit testing

To run the unit tests, execute one of the following commands:

*  `vendor/bin/phpunit tests/unit`
*  `vendor/bin/phpunit -c tests/unit/phpunit.xml.dist tests/unit`
*  `vendor/bin/phpunit -c tests/unit/phpunit.xml.dist --testsuite=unit-tests`

## Integration testing

To run the integration tests, execute one of the following commands:

*  `vendor/bin/phpunit -c tests/integration/phpunit.xml.dist tests/integration`
*  `vendor/bin/phpunit -c tests/integration/phpunit.xml.dist --testsuite=integration-tests`

## Acceptance testing

1. Before executing acceptance tests, you must set the Magento URL in the `phpunit` configuration file.
1. Copy the default `tests/acceptance/phpunit.xml` file (without the .dist suffix).
1. Change the `TESTS_BASE_URL` value to point to the Magento URL that you want to test.
1. To run the acceptance tests, execute one of the following commands:

> *  `vendor/bin/phpunit -c tests/acceptance/phpunit.xml tests/acceptance`

Reviewer 2: This unordered list needs to be indented so that it aligns under the preceding ordered list.

{:.bs-callout-info}
To run JS unit tests for GraphQL, you must have Node.js installed.

### Node.js

To install Node.js on your system, see the Node.js [documentation](https://nodejs.dev/learn/how-to-install-nodejs).

The following instructions are for MacOS systems:

1. Open a terminal and navigate to the `graphql-schema-compatibility/` directory.
1. Install project dependencies:

   ```bash
   npm install
   ```

1. Run unit tests or `jest`:

   ```bash
   npm run unit-test
   ```

The tests are inside `graphql-schema-compatibility/test/js/unit`.

The string schemas for testing are inside `dev/graphql_schemas`.

## Complexity score

The **complexity score** is a figure that indicates how difficult an upgrade from the current version to the new one might be. Lower numbers indicate easier upgrades.

{:.bs-callout-info}
Complexity scores range between 0 and 100.

This score is based on the results extracted from the analysis:

*  Number of issues identified
*  Severity of issues identified

The SUT calculates this score according to the following formula:

### Complexity score formula

> `Complexity Score = 2 * (# of errors) + 1 * (# of warnings)`

Reviewer 3: Above, the score is said to be in range (0, 100), which is normalized value.
While this formula is for absolute values, which can go beyond 100.

Please clarify with the team which one is currently implemented. Originally it was absolute value, and I have not seen it to be changed to normalized one yet.

Reviewer: Yes, it is an absolute value, we are planning to normalize it but it is still not implemented

Author: so how we could reflect this in the docs then?, should we change the absolute values mentioned in the formula, or modify the range, which is a normalized value?

Author: depending on when are we applying these changes

Author: @svera could you take a look at this thread too?, thanks!

Reviewer 4: @danidelcar right now the range can be from 0 to ∞. As Sandra said, we plan to implement a formula to return normalized values, but is not defined yet.

Author: corrected section, please check @paliarush & @svera

> ## Packaging

Reviewer 2: Have you confirmed whether external users can actually do packaging? This seems like an internal-only task.

Reviewer 3: Totally agree, there is no reason to have this section in public docs.

Author: will remove all packaging section then

> A package is essentially a directory containing something. It contains a package description which has a name and a version. The name and the version are used to identify the package.

Reviewer 2: This paragraph is awkward. It sounds like we're referring to a Composer package or Composer metapackage. 

https://getcomposer.org/doc/04-schema.md#type

If so, please use one of our official descriptions in the Magento Glossary:

- https://glossary.magento.com/composer-package/
- https://glossary.magento.com/metapackage/

Author: will remove all packaging section then

> It is possible to create a `sut.zip` file of the SUT to distribute it with partners as a composer package.

Reviewer 2: ```suggestion
It is possible to create a `sut.zip` file of the SUT to distribute it as a Composer package.
```

> In order to upload it to be able to distribute it, you need first to create the package:

Reviewer 2: ```suggestion
```

> 1. To package the SUT execute the following command:

Reviewer 2: ```suggestion
1. To package the SUT, execute the following command:
```

   ```bash
   bin/package VERSION`
   ```

### An example of a package

`bin/package 0.0.1`

> 1. This generates an `sut.zip` file that you can upload (or distribute with) to the [Magento Composer repository](https://repo.magento.com).

Reviewer 2: ```suggestion
1. This generates an `sut.zip` file that you can upload to the [Magento Composer repository](https://repo.magento.com).
```

Reviewer 2: As far as I know, only authorized employees can upload packages to repo.magento.com. Are you sure this belongs in our public docs?

Author: will remove all packaging section then

{:.bs-callout-info}
A repository is a package source. Composer will look in all your repositories to find the packages your SUT requires.

> If you need access to the [Magento Composer repository](https://repo.magento.com), submit a [Magento Support ticket]({{site.baseurl}}/cloud/trouble/trouble.html) to request permissions and an account.

Reviewer 2: Does support know we're directing users to them for assistance with SUT and repo.magento.com access?

Author: will remove all packaging section then

For more information about Composer, see the [Introduction to Magento Composer]({{site.baseurl}}/guides/v2.4/extension-dev-guide/intro/intro-composer.html) topic.

> ## Distribution

Reviewer 3: This section also does not make sense for external developers.

Once the SUT package is generated, you can upload it to the [Magento Composer repository](https://repo.magento.com).

In order to distribute SUT:

> 1. Go to the [Magento create repo](https://repo.magento.com/admin/packagist/web/ceeerelease/create) topic and create a new release with the following parameters:

Reviewer 2: ```suggestion
1. Go to the [Magento create repo](https://repo.magento.com/admin/packagist/web/ceeerelease/create) page and create a new release with the following parameters:
```

   *  Name: `magento/safe-upgrade-tool-VERSION`

> 1. Go to the [Magento upload repo](https://repo.magento.com/admin/upload_m2_version) topic to upload the `sut.zip` with the following parameters:

Reviewer 2: ```suggestion
1. Go to the [Magento upload repo](https://repo.magento.com/admin/upload_m2_version) page to upload the `sut.zip` file with the following parameters:
```

   *  Extension package: `sut.zip`
   *  Release: `magento/safe-upgrade-tool-VERSION`
   *  Edition: CE

> 1. Finally, go to the [Magento version repo](https://repo.magento.com/admin/packagist/web/version/list) topic to check that the new version has been created.

Reviewer 2: ```suggestion
1. Finally, go to the [Magento version repo](https://repo.magento.com/admin/packagist/web/version/list) page to check that the new version has been created.
```

Reviewer: We need to remove all the package section as highlighted by @paliarush

Author: will remove all packaging section then

Author: what about distribution @sgonzalezmangana?

Author: removed it, as per other comments
